metadata-ingestion/src/datahub/ingestion/source/gc/soft_deleted_entity_cleanup.py

```diff
@@ -157,68 +186,127 @@ def delete_soft_deleted_entity(self, urn: str) -> None:
             ]["created"]["time"] < (retention_time * 1000):
                 logger.debug(f"Hard deleting {urn}")
                 self.delete_entity(urn)
+            else:
+                self._increment_retained_count()
+
+    def _print_report(self) -> None:
+        time_taken = round(time.time() - self.last_print_time, 1)
+        # Print report every 2 minutes
+        if time_taken > 120:
+            self.last_print_time = time.time()
+            logger.info(f"\n{self.report.as_string()}")
+
+    def _process_futures(self, futures: Dict[Future, str]) -> Dict[Future, str]:
+        done, not_done = wait(futures, return_when=FIRST_COMPLETED)
+        futures = {future: urn for future, urn in futures.items() if future in not_done}
+
+        for future in done:
+            self._print_report()
+            if future.exception():
+                logger.error(
+                    f"Failed to delete entity {futures[future]}: {future.exception()}"
+                )
+                self.report.failure(
+                    f"Failed to delete entity {futures[future]}",
+                    exc=future.exception(),
+                )
```

Comment:
some redundancy here given `report.failure` already performs `logger.error`, am I wrong?

Reply:
I will check in follow up